pkg/interpreter/man/man1/osexec.1
.TH osexec "1" "January 9, 2021" "" "osexec manual"

.SH NAME
osexec - OScript interpreter

.SH SYNOPSIS
.B osexec
[--[no]-context] [--[no]-colors]
[-D|--define
.I name
]
.RS
[-S|--source
.I
type
] [-O|--old-version]
[-w|--warnings]
.br
[-s|--silent] [-v|--verbose]
[-p|--performance]
.br
[-V|--version] [-h|--help] [
.I file
]
.RE

.SH DESCRIPTION
.B
osexec is an interpreter executing OScript programs from source code.

.SH OPTIONS
.B
.IP "--[no]-context"
show near source as error context. defaults to true
.B
.IP "--[no]-colors"
enable colors in the terminal. default is auto
.B
.IP "-D|--define name"
define a named value for preprocessor
.B
.IP "-S|--source type"
source type is object, script or dump. defaults to script
.B
.IP "-O|--old-version"
expect an old version of OScript. defaults to false
.B
.IP "-w, --warnings"
consider warnings as failures too
.B
.IP "-s, --silent"
suppress output
.B
.IP "-v|--verbose"
print error stacktrace
.B
.IP "-p|--performance"
print parsing timing
.B
.IP "-V, --version"
print version number
.B
.IP "-h, --help"
print usage instructions

.RE
If no file name is provided, standard input will be read. If no source type
is provided, it will be inferred from the file extension: ".os" -> object,
".e|lxe" -> script, ".osx" -> dump. The source type object will enable the
new OScript language and source type dump the old one by default.

.SH EXAMPLES
.B
.IP "echo 'echo(""foo"")' | osexec -S script"
Executes a script from the standard input.
.B
.IP "osexec -w foo.e"
Executes a script from a file.

.SH BUGS
osexec's bug tracker is located at <https://github.com/prantlf/oscript-parser/issues>

.SH DOCUMENTATION
See the website for documentation <https://github.com/prantlf/oscript-parser/tree/master/pkg/interpreter#readme>

.SH AUTHOR
Currently maintained by Ferdinand Prantl <prantlf@gmail.com>


CHANGELOG.md
# Changelog

## 0.2.6

Recognise lxe as a script extension

### Interpreter 0.2.0

Add initial support for File.

### Interpreter 0.3.0

Temporarily remove functions `eof` and `seek`. The `fs-ext` extension with a native library does not run in the latest VS Code

**BREAKING CHANGE**: If you used `eof` or `seek`, your code will not work any more. You have to remove calls to those functions.

### Walker 0.1.1

Fix typings for `findNodeAround` and `findNodeAroundWithAncestors`.

## 0.2.5

### Parser

* Allow dereferencing `this` without the `this` keyword (chain the dot operators).
* Correct typings for `SliceExpression`.
* Nest the binary expressions according to the binary operator precedence.

### Walker 0.1.0

If the recursive walker fails, include the latest visited node in the error.

### Interpreter 0.1.0

* Ignore the imaginary operator `^^`, recognise the existing one `||`.
* Fix built-in object method lookup.
* Set default values to declared variables.
* Introduce built-in objects `Assoc`, `List` and `Regex`.
* Add runtime type checking to the built-in functions and object methods.
* Include the last interpreted node in the runtime error report.
* Inline the preparation of AST from an extra preprocessing step to the interpretation phase.

**BREAKING CHANGE**: The second parameters of `interpret` is an object with options. The parameter `globals` has been moved to a property in the `options` object.

```js
// old syntax
interpret(ast, { count: 1 })
// new syntax
interpret(ast, { globals: { count: 1 } })
```

## 0.2.4

### Parser

Remove the operator `~=` that does not exist.

### Walker 0.0.3

Skip omitted nodes `for.init`, `slice.start` and `slice.end` expressions during walking.

### Interpreter 0.0.1

Initial release.

## 0.2.3

Fix parsing of an incomplete `switch` statement.

## 0.2.2

### Parser

Fix parsing of the member expression without the explicit `this` object.

### Walker 0.0.2

* Recognize `ObjectExpression` as `AssocExpression` for compatibility.
* Fix walking of `consequent` and `alternate` of `ConditionalExpression`.

## 0.2.1

Fix parsing of the `switch` statement.

## 0.2.0

### New Features

* Allow storing raw content not only for literals, but for identifiers too.
* Let warning fail the syntax check too, if requested by `-w|--warnings`.
* Use colours in the console output.
* Print error context for parsing failures, for exaample:

    $ oslint example.os

    example.os failed with 1 error and 0 warnings
    example.os:7:28: error: modifier, type, function, script or end expected near 'public'
     5｜
     6｜ public object Document inherits CORE::Node
     7｜  override Boolean fEnabled TRUE
      ｜                            ~~~~
     8｜
     9｜  // Gets a livelink document

### Fixed Bugs

* Correct parsed node and error locations.
* Correct typings for AST nodes.

### Walker 0.0.1

Initial release.

## 0.1.0

Include a new property `warnings` in the parser output or error, which will contain an array of warnings. A warning is a mistake in the source code, which is not fatal (and will be compiled and executed properly), but which is against the modern language specification. Warn  about:

* A line break in a single-line string
* A backslash not followed by a whitespace
* A `ifdef` or `ifndef` preprocessor directive without a name identifier following it
* An object declared with other modifier than `public`
* A preprocessor directive followed by non-whitespace characters
* A semicolon following feature, function or script declaration
* A missing line break or semicolon after an empty c-like for statement before the end keyword

## 0.0.1

Initial release.

* Command-line tools `osparse` and `oslint`
* Functions `parseText`, `parseTokens`, `tokenize` and `startTokenization`
* Typescript typings


pkg/interpreter/src/osexec.js
import { performance } from 'perf_hooks'
import { options as colors, bold, white, yellow } from 'colorette'
import { parseText } from 'oscript-parser'
import { interpret } from 'oscript-interpreter'
import {
  version, matchArgument, slurpStdin, readSource, printError, printWarnings
} from '../../osparse/console'

const args = process.argv
let considerWarnings = false
let context = true
let silent = false
let verbose = false
let measure = false
const defines = {}
const options = { defines, locations: true, ranges: true }
let source, sourceType

function usage () {
  console.log(`Executes OScript programs.

${bold(yellow('Usage:'))} ${bold(white('osexec [option...] [file]'))}

${bold(yellow('Options:'))}
  --[no]-context          show near source as error context. defaults to true
  --[no]-colors           enable colors in the terminal. default is auto
  -D|--define <name>      define a named value for preprocessor
  -S|--source <type>      source type is object, script (default) or dump
  -O|--old-version        expect an old version of OScript. defaults to false
  -w|--warnings           consider warnings as failures too
  -s|--silent             suppress error output
  -v|--verbose            print error stacktrace
  -p|--performance        print parsing timing
  -V|--version            print version number
  -h|--help               print usage instructions

If no file name is provided, standard input will be read. If no source type
is provided, it will be inferred from the file extension: ".os" -> object,
".e|lxe" -> script, ".osx" -> dump. The source type object will enable the
new OScript language and source type dump the old one by default.

${bold(yellow('Examples:'))}
  echo 'echo("foo")' | osexec -S script
  osexec -w foo.e`)
  process.exit(0)
}

if (!args.length) usage()

for (let i = 2, l = args.length; i < l; ++i) {
  const arg = args[i]
  const match = matchArgument(arg)
  if (match) {
    const flag = match[2]
    switch (flag) {
      case 'w': case 'warnings':
        considerWarnings = options.warnings = true
        continue
      case 'context':
        context = match[1] !== 'no'
        continue
      case 'colors':
        colors.enabled = match[1] !== 'no'
        continue
      case 's': case 'silent':
        silent = true
        continue
      case 'v': case 'verbose':
        verbose = true
        continue
      case 'p': case 'performance':
        measure = true
        continue
      case 'D': case 'define':
        defines[args[++i]] = true
        continue
      case 'S': case 'source':
        sourceType = options.sourceType = args[++i]
        continue
      case 'O': case 'old-version':
        options.oldVersion = true
        continue
      case 'V': case 'version':
        console.log(version)
        process.exit(0)
        break
      case 'h': case 'help':
        usage()
    }
    if (!silent) console.error(`Unknown option: "${match[0]}".`)
    process.exit(2)
  }
  if (source) {
    if (!silent) console.error('More than one file supplied.')
    process.exit(2)
  }
  source = readSource(arg, sourceType, options)
}

function run () {
  let start, ast
  try {
    options.sourceFile = source.name
    start = measure && performance.now()
    ast = parseText(source.code, options)
    const end = start && performance.now()
    let end2
    if (considerWarnings && ast.warnings.length) {
      if (!silent) printWarnings(process.stderr, source, ast.warnings, { context })
      process.exitCode = 1
    } else {
      interpret(ast, options)
      end2 = start && performance.now()
    }
    if (start) {
      const time = end2
        ? `parsing ${Math.round(end - start)}, executing ${Math.round(end2 - end)})`
        : `${Math.round(end - start)}ms`
      process.stderr.write(`time: ${time}\n`)
    }
  } catch (error) {
    if (!silent) {
      if (!error.warnings) error.warnings = ast && ast.warnings
      printError(process.stderr, source, start, error, { context, verbose })
    }
    process.exitCode = 1
  }
}

if (source) {
  run()
} else {
  slurpStdin(result => {
    source = result
    run()
  })
}
